service/src/main/java/aero/minova/cas/MultipleLdapDomainsAuthenticationProvider.java

```diff
@@ -0,0 +1,66 @@
+package aero.minova.cas;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.springframework.security.authentication.AuthenticationProvider;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.ldap.authentication.ad.ActiveDirectoryLdapAuthenticationProvider;
+import org.springframework.security.ldap.userdetails.UserDetailsContextMapper;
+
+public class MultipleLdapDomainsAuthenticationProvider implements AuthenticationProvider {
+
+	List<ActiveDirectoryLdapAuthenticationProvider> providers = new ArrayList<>();
+
+	public MultipleLdapDomainsAuthenticationProvider(String domains, String ldapServerAddresses, UserDetailsContextMapper userDetailsContextMapper) {
+
+		List<String> domainList = Arrays.asList(domains.split(SecurityConfig.MULTIPLE_LDAP_CONFIGURATIONS_SEPERATOR));
+		List<String> addressList = Arrays.asList(ldapServerAddresses.split(SecurityConfig.MULTIPLE_LDAP_CONFIGURATIONS_SEPERATOR));
+
+		for (int i = 0; i < domainList.size(); i++) {
+
+			String address = addressList.size() == 1 ? addressList.get(0) : addressList.get(i);
+
+			ActiveDirectoryLdapAuthenticationProvider provider = new ActiveDirectoryLdapAuthenticationProvider(domainList.get(i), address);
+			provider.setConvertSubErrorCodesToExceptions(true);
+			provider.setUserDetailsContextMapper(userDetailsContextMapper);
+
+			providers.add(provider);
+		}
+	}
+
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+
+		for (ActiveDirectoryLdapAuthenticationProvider provider : providers) {
+
+			try {
+				Authentication authenticate = provider.authenticate(authentication);
+				if (authenticate != null) {
```

Comment:
Ich glaube, dass auch `authenticate.isAuthenticated()` geprüft werden muss, bin mir aber auch nicht sicher: https://docs.spring.io/spring-security/site/docs/4.0.x/apidocs/org/springframework/security/core/Authentication.html#isAuthenticated--

`setAuthenticated` kann von den AuthenticationProvider anscheinend benutzt werden, um Authorisierungsanfragen als nicht vertrauenswürdig zu markieren: https://docs.spring.io/spring-security/site/docs/4.0.x/apidocs/org/springframework/security/core/Authentication.html#setAuthenticated-boolean-

Reply:
Ich hatte mir gedacht, in dem Fall das genau eine Domäne und Adresse konfiguriert ist sollte es sich genauso verhalten wie vorher.
In diesem Fall würde ja auch nur `authenticate`  von dem einen Provider aufgerufen werden, alles andere macht die `ActiveDirectoryLdapAuthenticationProvider`-Implementierung, die wir nutzen.

Reply:
Ich weiß leider nicht, ob es bei 2 Servern zu Problemen führt, wenn der erste Server den Nutzer nicht kennt,
dies aber der zweite Server kennt.

Wir können es erstmal so lassen und dann anpassen,
wenn wir merken, dass es Probleme verursacht.